flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaEvolutionUtil.java

```diff
@@ -62,4 +72,193 @@ public static int[] createIndexMapping(
+        for (int i = 0; i < tableKeyCount; i++) {
+            int fieldId = tableKeyFieldIdList.get(tableProjection[i]);
+            int dataFieldIndex = dataKeyFieldIdList.indexOf(fieldId);
+            indexMapping[i] = Ints.indexOf(dataKeyProjection, dataFieldIndex);
+        }
+        if (tableProjection.length >= tableKeyCount + 2) {
+            // seq and value kind
+            for (int i = tableKeyCount; i < tableKeyCount + 2; i++) {
+                indexMapping[i] = i + dataKeyCount - tableKeyCount;
+            }
+
+            int[] dataValueProjection =
+                    Arrays.copyOfRange(dataProjection, dataKeyCount + 2, dataProjection.length);
+            for (int i = 0; i < dataValueProjection.length; i++) {
+                dataValueProjection[i] = dataValueProjection[i] - dataKeyFields.size() - 2;
+            }
+            List<Integer> tableValueFieldIdList =
+                    tableValueFields.stream().map(DataField::id).collect(Collectors.toList());
+            List<Integer> dataValueFieldIdList =
+                    dataValueFields.stream().map(DataField::id).collect(Collectors.toList());
+            for (int i = tableKeyCount + 2; i < tableProjection.length; i++) {
+                int fieldId =
+                        tableValueFieldIdList.get(tableProjection[i] - tableKeyFields.size() - 2);
+                int dataFieldIndex = dataValueFieldIdList.indexOf(fieldId);
+                int dataValueIndex = Ints.indexOf(dataValueProjection, dataFieldIndex);
+                indexMapping[i] =
+                        dataValueIndex < 0 ? dataValueIndex : dataValueIndex + dataKeyCount + 2;
+            }
+        }
+
+        for (int i = 0; i < indexMapping.length; i++) {
+            if (indexMapping[i] != i) {
+                return indexMapping;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Create data projection from table projection.
+     *
+     * @param tableFields the fields of table
+     * @param dataFields the fields of underlying data
+     * @param tableProjection the projection of table
+     * @return the projection of data
+     */
+    public static int[][] createDataProjection(
+            List<DataField> tableFields, List<DataField> dataFields, int[][] tableProjection) {
+        List<Integer> tableFieldIdList =
+                tableFields.stream().map(DataField::id).collect(Collectors.toList());
+        List<Integer> dataFieldIdList =
+                dataFields.stream().map(DataField::id).collect(Collectors.toList());
+        return Arrays.stream(tableProjection)
+                .map(p -> Arrays.copyOf(p, p.length))
+                .peek(
+                        p -> {
+                            int fieldId = tableFieldIdList.get(p[0]);
+                            p[0] = dataFieldIdList.indexOf(fieldId);
+                        })
+                .filter(p -> p[0] >= 0)
```

Comment:
Why filter negative index?
For example, `tableProjection` is {fieldId-0, fieldId-1, fieldId-2}. `fieldId-2` can not be found in data fields.
But we should produce a row with three columns? The third column should be null?

Reply:
I may have confused it with `createIndexMapping`